plugins/auto-sizes/hooks.php

```diff
@@ -139,79 +139,145 @@ function auto_sizes_get_width( string $layout_width, int $image_width ): string
 }
 
 /**
- * Filter the sizes attribute for images to improve the default calculation.
+ * Primes attachment into the cache with a single database query.
  *
- * @since 1.1.0
+ * @since n.e.x.t
  *
- * @param string                                                   $content      The block content about to be rendered.
- * @param array{ attrs?: array{ align?: string, width?: string } } $parsed_block The parsed block.
- * @return string The updated block content.
+ * @param string|mixed $content The HTML content.
+ * @return string The HTML content.
  */
-function auto_sizes_filter_image_tag( string $content, array $parsed_block ): string {
+function auto_sizes_prime_attachment_caches( $content ): string {
+	if ( ! is_string( $content ) ) {
+		return '';
+	}
+
 	$processor = new WP_HTML_Tag_Processor( $content );
-	$has_image = $processor->next_tag( array( 'tag_name' => 'img' ) );
 
-	// Only update the markup if an image is found.
-	if ( $has_image ) {
-		$processor->set_attribute( 'data-needs-sizes-update', true );
-		if ( isset( $parsed_block['attrs']['align'] ) ) {
-			$processor->set_attribute( 'data-align', $parsed_block['attrs']['align'] );
+	$images = array();
+	while ( $processor->next_tag( array( 'tag_name' => 'IMG' ) ) ) {
```

Comment:
This seems like a good use of the `WP_HTML_Tag_Processor`, and is fine for a first pass at this functionality, but it would be good to do some profiling of this method against the `preg_match_all` approach that [is currently in use in Core](https://github.com/WordPress/wordpress-develop/blob/61b9859728b31130274b31bde9c6d73ee915fa03/src/wp-includes/media.php#L1868-L1870).

If we eventually want to land this code in WP, we'll want to move this priming logic from `wp_filter_content_tags` rather than duplicate it in two places, at which time we should be mindful of any performance differences between the current logic and this slightly updated version that makes use of the `WP_HTML_Tag_Processor`.

Reply:
> This seems like a good use of the `WP_HTML_Tag_Processor`, and is fine for a first pass at this functionality, but it would be good to do some profiling of this method against the `preg_match_all` approach that [is currently in use in Core](https://github.com/WordPress/wordpress-develop/blob/61b9859728b31130274b31bde9c6d73ee915fa03/src/wp-includes/media.php#L1868-L1870).

Haven't we already debated this at length, with the conclusion that using `WP_HTML_Tag_Processor` is preferred for correctness and that the practical performance impact is inconsequential?

Reply:
In previous discussion in #1471 we preferred to use `WP_HTML_Tag_Processor`.

Reply:
> Haven't we already debated this at length, with the conclusion that using WP_HTML_Tag_Processor is preferred for correctness and that the practical performance impact is inconsequential?

I keep seeing this same question asked in almost every PR where `WP_HTML_Tag_Processor` is used in place of other approaches, so if a consensus conclusion was reached, I was not aware of it. I'm more than happy for this to be the decision though.